api/src/knowledge.py

```diff
@@ -0,0 +1,342 @@
+
+from __future__ import annotations
```

Comment:
Our boxytas code and grounded causal recommended was virtually copied here. I know we could split into separate files, but I've left that for later for now.